src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java

```diff
@@ -79,4 +94,16 @@ private Book getBookByIdOrThrowException(Long id) {
                         localeService.getMessage("exception.notfound.book") + id
                 ));
     }
+
+    private void validateBookCategoriesExistence(Set<Category> categories) {
+        categories.forEach(i -> {
+            Long categoryId = i.getId();
+            boolean isCategoryExists = categoryRepository.existsById(categoryId);
+            if (!isCategoryExists) {
+                throw new EntityNotFoundException(
+                        localeService.getMessage("exception.notfound.category") + categoryId
+                );
+            }
+        });
+    }
```

Comment:
You make N requests to check if categories exists, investigate how to make it using less db requests (maybe you can fetch all categories that exists by id and compare size of retrieved list to number of categories in request)